### PR TITLE
Fixed issue #3665

### LINKED ISSUE #3665: Project deletion creates a "zombie" project in the backend
Destroying a Project ends up in a project meta in the backend for that project making to the undelete command impossible to recover the original project.

It only happens if the project have at least 2 repositories and one is linking to the other (using path element), for example:

```xml
<project name="home:user:project">
  <title/>
  <description/>
  <person userid="user" role="bugowner"/>
  <person userid="user" role="maintainer"/>
  <build>
    <enable/>
  </build>
  <publish>
    <disable/>
  </publish>
  <debuginfo>
    <enable/>
  </debuginfo>
  <repository name="images" block="local">
    <path project="OBS:Server:2.8" repository="SLE_12_SP2"/>
    <path project="home:user:project" repository="SLE_12"/>
    <path project="openSUSE:Tools" repository="SLE_12"/>
    <arch>x86_64</arch>
  </repository>
  <repository name="SLE_12">
    <path project="SUSE:SLE-12-SP2:Update" repository="standard"/>
    <arch>x86_64</arch>
  </repository>
</project>
```

This issue closes #3631 because the cause root is the deletion of the project not the undelete command.

### PR DESCRIPTION
**Pull request by mdeniz**
Delete project on backend at the end of the callbacks cycle, first cleanup everything (linking repos, linking projects, request, etc), then remove the database record and then the backend entry.

This avoids to store the project in the backend after removing it in the cleanup of the Project model.

It fixes #3665.

### CODE REVIEW

File: src/api/app/models/project.rb [hunk 1]
```diff
@@ -31,6 +31,7 @@ class ForbiddenError < APIException
   end
 
   before_destroy :cleanup_before_destroy
+  after_destroy :delete_on_backend
```
Comment (hennevogel) [inline src/api/app/models/project.rb]: Maybe [after_destroy_commit](http://guides.rubyonrails.org/active_record_callbacks.html#transaction-callbacks) would be even better

**Comment by adrianschroeter** [comment]
but this does not avoid the additional PUT, right?

So we would restore a project meta with the "deleted" repository?

File: src/api/app/models/project.rb (hunk 1 above)
Comment (mdeniz) [inline src/api/app/models/project.rb]: Sure, is even better

**Comment by mdeniz** [comment]
@adrianschroeter yes, that's the only case. But right now I don't know how to differentiate between a project.destroy that ends up breaking the linking repo, and a normal repo.destroy that also do the same (having in mind that we have the same repos setup in only one project, one linking to the other).

So, If I skip the store of the project, it will be a problem on a normal repository.destroy, it won't store the change (I'm not talking about what happens in the repository controller # destroy, I mean at model level).

File: src/api/app/models/project.rb (hunk 1 above)
Comment (mdeniz) [inline src/api/app/models/project.rb]: On the other hand the `delete_on_backend` rescues any exception related to backend communications...

**Comment by mdeniz** [comment]
To clarify:

If we skip the store of the project when updating the linking_repos it will work because a normal removing of the repository also stores the meta of the project in the controller.

But could happen that if we remove a repo in other places of the code without storing the project meta then the deleted/deleted is not assigned to the linking_repo

File: src/api/app/models/project.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/app/models/project.rb]: Only the *doesn't exist* case...

**Comment by mdeniz** [comment]
@hennevogel re-review, 👍

File: src/api/spec/models/project_spec.rb [hunk 2]
```diff
@@ -453,4 +452,20 @@
...
+      let(:images_repository) { create(:repository, name: "images", project: project) }
+      let(:apache_repository) { create(:repository, name: "Apache", project: project) }
+      let!(:path_element) { create(:path_element, parent_id: images_repository.id, repository_id: apache_repository.id, position: 1)}
+
+      before do
+        login admin_user
+        project.destroy!
+      end
+
+      it { expect(Project.deleted?(project.name)).to be_truthy }
```
Comment (hennevogel) [inline src/api/spec/models/project_spec.rb]: Hm. This was always the case right?

Comment (mdeniz) [inline src/api/spec/models/project_spec.rb]: No, with this setup...
In this case with having 2 repos, one linking to the other in the same project, the delete operation is executed in the backend and right after that a put with the meta was performed, so the backend had a project already after deleting it.

That's the cause of the issue, this test is only for avoiding the regression.

**Comment by mdeniz** [comment]
So, @adrianschroeter what should we do...?? fix this and open another issue about the undeletion of the project that has been touched after removing because they have a similar setup of repos?

**Pull request merged**